java/libphonenumber/src/com/google/i18n/phonenumbers/MultiFileMetadataSourceImpl.java

```diff
@@ -73,91 +67,81 @@ public MultiFileMetadataSourceImpl(MetadataLoader metadataLoader) {
-      if (!regionToMetadataMap.containsKey(regionCode)) {
-        // The regionCode here will be valid and won't be '001', so we don't need to worry about
-        // what to pass in for the country calling code.
-        loadMetadataFromFile(regionCode, 0);
-      }
-    }
-    return regionToMetadataMap.get(regionCode);
+    PhoneMetadata metadata = regionToMetadataMap.get(regionCode);
+    return (metadata != null) ? metadata : loadMetadataFromFile(
+        regionCode, regionToMetadataMap, filePrefix, metadataLoader);
   }
 
   @Override
   public PhoneMetadata getMetadataForNonGeographicalRegion(int countryCallingCode) {
-    synchronized (countryCodeToNonGeographicalMetadataMap) {
-      if (!countryCodeToNonGeographicalMetadataMap.containsKey(countryCallingCode)) {
-        List<String> regionCodes =
-            CountryCodeToRegionCodeMap.getCountryCodeToRegionCodeMap().get(countryCallingCode);
-        // We can assume that if the country calling code maps to the non-geo entity region code
-        // then that's the only region code it maps to.
-        if (regionCodes.size() == 1 &&
-            PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY.equals(regionCodes.get(0))) {
-          loadMetadataFromFile(PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY, countryCallingCode);
-        }
-      }
+    PhoneMetadata metadata = countryCodeToNonGeographicalMetadataMap.get(countryCallingCode);
+    if (metadata != null) {
+      return metadata;
     }
-    return countryCodeToNonGeographicalMetadataMap.get(countryCallingCode);
+    if (isNonGeographicalCountryCallingCode(countryCallingCode)) {
+      return loadMetadataFromFile(
+          countryCallingCode, countryCodeToNonGeographicalMetadataMap, filePrefix, metadataLoader);
+    }
+    return null;
+  }
+
+  private boolean isNonGeographicalCountryCallingCode(int countryCallingCode) {
+    List<String> regionCodes =
+        CountryCodeToRegionCodeMap.getCountryCodeToRegionCodeMap().get(countryCallingCode);
+    // A country calling code is non-geographic if it only maps to the non-geographical region code.
+    return (regionCodes.size() == 1 &&
+        PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY.equals(regionCodes.get(0)));
   }
 
+  // @param key  The geographical region code or non-geographical region's country calling code.
   // @VisibleForTesting
-  void loadMetadataFromFile(String regionCode, int countryCallingCode) {
-    boolean isNonGeoRegion = PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY.equals(regionCode);
-    String fileName = filePrefix + "_" +
-        (isNonGeoRegion ? String.valueOf(countryCallingCode) : regionCode);
+  static <T> PhoneMetadata loadMetadataFromFile(
+      T key, ConcurrentHashMap<T, PhoneMetadata> map, String filePrefix,
+      MetadataLoader metadataLoader) {
+    // We assume key.toString() is well-defined.
+    String fileName = filePrefix + "_" + key;
     InputStream source = metadataLoader.loadMetadata(fileName);
```

Comment:
This API is now badly named (it's not a "loader" and doesn't load the data). It's a stream provider.